src/core/dense_set.cc

```diff
@@ -386,6 +386,57 @@ void DenseSet::Reserve(size_t sz) {
   }
 }
 
+void DenseSet::ShrinkBucket(size_t bucket_idx) {
+  // Take the entire bucket to avoid infinite loop when new_bid == bucket_idx
+  DensePtr bucket = entries_[bucket_idx];
+  entries_[bucket_idx].Reset();
+  --num_used_buckets_;
+
+  // Process the taken bucket chain
+  while (!bucket.IsEmpty()) {
+    // Pop front from local chain
+    DensePtr dptr = bucket;
+    bucket = bucket.IsObject() ? DensePtr{} : bucket.AsLink()->next;
+
+    void* obj = dptr.GetObject();
+    bool has_ttl = dptr.HasTtl();
+
+    // Free link unconditionally - PushFront will create new one if needed
+    if (dptr.IsLink()) {
+      FreeLink(dptr.AsLink());
+    }
+
+    if (has_ttl && ObjExpireTime(obj) <= time_now_) {
+      ObjDelete(obj, true);
+      --size_;
+      continue;
+    }
+
+    uint32_t new_bid = BucketId(obj, 0);
```

Comment:
based on the hash functon, it looks like `2*i`  and `i` are mapped to -> `i`. Not sure if its worth dcheck'ing or commenting, because the code looks more generic than it already is